### PR TITLE
openstack-virtual-ip: New resource agent for virtual IP support in OpenStack.

### PR DESCRIPTION
**Pull request by bdperkin**
- New resource agent for virtual IP support in OpenStack.

### CODE REVIEW

File: heartbeat/openstack-virtual-ip [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+		exit $OCF_SUCCESS
+		;;
+esac
+
+if ! ocf_is_root; then
+	ocf_log err "You must be root for $__OCF_ACTION operation."
+	exit $OCF_ERR_PERM
+fi
+
+osvip_validate
```
Comment (oalbrigt) [inline heartbeat/openstack-virtual-ip:260]: You'll have to use exit in the validate function, or the result will be ignored when action isnt validate-all.

File: heartbeat/openstack-virtual-ip [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+		return $OCF_ERR_GENERIC
+	fi
+
+	osvip_monitor
+	if [ $? != $OCF_NOT_RUNNING ]; then
+		ocf_log error "Couldn't unset IP address $OCF_RESKEY_ip."
+		return $OCF_ERR_GENERIC
+	fi
+
+	ocf_log info "Successfully brought unset $OCF_RESKEY_ip"
```
Comment (oalbrigt) [inline heartbeat/openstack-virtual-ip:206]: This sounds a bit strange to me, and I think detached is a better word for unset in the situation.

**Review by oalbrigt — CHANGES_REQUESTED** [review]

File: heartbeat/openstack-virtual-ip (hunk 1 above)
Comment (bdperkin) [reply heartbeat/openstack-virtual-ip]: Fixed with commit 61aebd2c482b0052b2865fba01acc11bc0fc5fab:
```
-osvip_validate
-
 case $__OCF_ACTION in
 	start)
-		osvip_start;;
+		osvip_validate
+		osvip_start
+		;;
 	stop)
 		osvip_stop;;
 	monitor)
 		osvip_monitor;;
 	validate-all)
-		exit $?;;
+		osvip_validate;;
```

File: heartbeat/openstack-virtual-ip (hunk 2 above)
Comment (bdperkin) [reply heartbeat/openstack-virtual-ip]: Fixed with commit 61aebd2c482b0052b2865fba01acc11bc0fc5fab:
```
-	ocf_log info "Successfully brought unset $OCF_RESKEY_ip"
+	ocf_log info "Successfully brought down $OCF_RESKEY_ip"
```

File: heartbeat/openstack-virtual-ip (hunk 1 above)
Comment (bdperkin) [reply heartbeat/openstack-virtual-ip]: I actually changed this again (once I realize I broke the agent with the change).  Now, osvip_validate is still removed from the main and has been added to the beginning of the monitor function.  The reason for this is that start|stop|monitor all require the openstack CLI and the resource file that the validation checks.  The sourcing of the resource file also happens in the validation function, so by adding validation to monitoring and all the appropriate functions call monitoring, the proper checks and environment variables are now set.

**Review by oalbrigt — APPROVED** [review]

**Comment by oalbrigt** [comment]
LGTM. Thanks.

**Pull request merged**